README.md
# 经典水果机抽奖
[fruit-slot.js](https://github.com/ufologist/fruit-slot/blob/master/src/fruit-slot.js)@[2.1.0](https://github.com/ufologist/fruit-slot/blob/master/CHANGELOG.md)

![fruit-slot-classic](http://ufologist.github.io/fruit-slot/fruit-slot-classic.png)

用于9宫格/12宫格等格子似的抽奖表现形式.

fork 后去除了[@twlk28 版本](https://github.com/duowan/lottery)中依赖的第三方库(例如: jQuery). 现在是无依赖([UMD](https://github.com/umdjs/umd))版, 适合作为一个很小的功能模块集成到项目中全局使用或作为 AMD/CMD 模块使用.

## Usage
组件本身只提供了抽奖的逻辑处理, 具体的抽奖样式, 可以自己任意发挥, 具体使用方法请参考[演示示例](http://ufologist.github.io/fruit-slot)

## Api

### 构造参数
```javascript
var fruitSlot = new FruitSlot({
    el: '#J_lottery1', // 抽奖容器元素，默认值 body，页面有多个抽奖动画时，需指定
    lottoItem: '[data-lotto-item]', // 用来指定抽奖容器内的奖品元素，默认值 [data-lotto-item]
    activeClass: 'lotto__item--hover', // 执行抽奖动画时切换的样式名，默认值 lotto__item--hover
    lottoArr: lottoArr // 手工指定奖品元素的顺序, 例如想将默认的顺时针元素排列改为逆时针排列, 只要你控制了元素的顺序就控制了抽奖动画执行的顺序
});
```

### 方法
```javascript
/**
 * 开始抽奖
 */
fruitSlot.start();

/**
 * 停止抽奖
 * 
 * @param  {number}   index 表示动画停在哪个奖品序号，奖品序号从0开始以左上角顺时针排列
 * @param  {Function} cb    动画停止后执行的回调函数
 * @param  {*}        ctx   回调函数的上下文
 */
fruitSlot.stop(index, cb, ctx);
```

## 浏览器兼容性
要使用该组件需要浏览器支持以下特性
* Array.forEach ES5(IE9+才支持)
* querySelector / querySelectorAll
* classList
* getBoundingClientRect

如果你需要兼容老版本的浏览器, 请使用 [@twlk28 的版本](https://github.com/duowan/lottery)

## 可供参考的其他项目
* [roulette.js](https://github.com/akira-kuriyama/roulette.js) Rotate image like a roulette game 用于老虎机
* [druphliu/Lottery](https://github.com/druphliu/Lottery) 经典水果机
* [chenyueling/lottery](https://github.com/chenyueling/lottery) 数字老虎机

    通过 repeat-y/background-position-y 来实现滚动
    因此可以使用 transition 来实现滚动的效果
    ```javascript
    // http://easings.net/zh-cn#easeInOutCirc
    $('.num').eq(0).css({
      'transition': 'background-position-y 6s 0s cubic-bezier(0.785, 0.135, 0.15, 0.86)',
      'background-position-y': '12792px'
    });
    $('.num').eq(1).css({
      'transition': 'background-position-y 9s 300ms cubic-bezier(0.785, 0.135, 0.15, 0.86)',
      'background-position-y': '11726px'
    });
    $('.num').eq(2).css({
      'transition': 'background-position-y 12s 600ms cubic-bezier(0.785, 0.135, 0.15, 0.86)',
      'background-position-y': '11512px'
    });
    ```

* [比较常用的几种抽奖效果](https://github.com/areyouse7en/lottery) 跑马灯 扑克牌翻转
* [大转盘](https://github.com/LucyLiuluxi/lottery)
* [移动端抽奖插件[大转盘，老虎机，刮刮卡]](https://github.com/TOP-Chao/lottery)
* [Github](https://github.com/search?utf8=%E2%9C%93&q=lottery)

HISTORY.md
# History

---

## 2.1.0 2016/1/21
构造函数新增 `lottoArr` 参数用于手工指定奖品元素的顺序, 例如想将默认的顺时针元素排列改为逆时针排列, 只要你控制了元素的顺序就控制了抽奖动画执行的顺序

## 2.0.0 2016/1/20
基于[@twlk28 的版本](https://github.com/duowan/lottery)去除了依赖的第三方库(例如: jQuery)

## 1.0.0

`new` /lottery First version.


src/lottery.js
// [UMD](https://github.com/umdjs/umd)
(function webpackUniversalModuleDefinition(root, factory) {
    if(typeof exports === 'object' && typeof module === 'object') {
        module.exports = factory();
    } else if(typeof define === 'function' && define.amd) {
        define([], factory);
    } else if(typeof exports === 'object') {
        exports['Lottery'] = factory();
    } else {
        root['Lottery'] = factory();
    }
})(this, function() {
    var STATUS_STARTING = 1, STATUS_STOPPING = 2, STATUS_STOPPED = 3,
        SPEED_HIGH = 50, SPEED_LOW = 300;

    var defaults = {
        el: "body",
        lottoItem: "[data-lotto-item]",
        activeClass: "lotto__item--hover"
    };

    // utils
    // http://youmightnotneedjquery.com/#extend
    function extend(out) {
        out = out || {};

        for (var i = 1; i < arguments.length; i++) {
            if (!arguments[i])
                continue;

            for (var key in arguments[i]) {
                if (arguments[i].hasOwnProperty(key))
                    out[key] = arguments[i][key];
            }
        }

        return out;
    }
    function _createCycleArr(lottoElements) {
        var arrSrc = [],arrOrder = [],arrDst = [],topMin = 9999,topMax = 0,leftMin = 9999,leftMax = 0,_arrTop = [],_arrBottom = [],_arrLeft = [], _arrRight = [];

        for (var i = 0, length = lottoElements.length; i < length; i++) {
            var lottoElement = lottoElements[i];
            var obj = lottoElement.getBoundingClientRect()
            obj.elem = lottoElement
            arrSrc.push(obj)
        }

        arrSrc.forEach(function(e) {
            topMin = e.top < topMin ? e.top : topMin;
            topMax = e.top > topMax ? e.top : topMax;
            leftMin = e.left < leftMin ? e.left : leftMin;
            leftMax = e.left > leftMax ? e.left : leftMax;
        })
        arrSrc.forEach(function(e) {
            e.top === topMin && _arrTop.push(e)
            e.top === topMax && _arrBottom.push(e)
            e.left === leftMin && _arrLeft.push(e)
            e.left === leftMax && _arrRight.push(e)
        })
        _arrTop.sort(function(a, b) {
            return a.left - b.left
        })
        _arrRight.sort(function(a, b) {
            return a.top - b.top
        })
        _arrBottom.sort(function(a, b) {
            return b.left - a.left
        })
        _arrLeft.sort(function(a, b) {
            return b.top - a.top
        })
        _arrTop = _arrTop.slice(0, _arrTop.length-1)
        _arrRight = _arrRight.slice(0, _arrRight.length-1)
        _arrBottom = _arrBottom.slice(0, _arrBottom.length-1)
        _arrLeft = _arrLeft.slice(0, _arrLeft.length-1)
        arrOrder = _arrTop.concat(_arrRight.concat(_arrBottom.concat(_arrLeft)))
        arrOrder.forEach(function(e) {
            arrDst.push(e.elem)
        })
        return arrDst
    }

    function Lottery(opts) {
        // 0. setup config
        var options = this.options = extend({}, defaults, opts)
        var arr = document.querySelector(options.el).querySelectorAll(options.lottoItem),
            lottoArr = _createCycleArr(arr), //格子们顺时针排好队
            activeClass = options.activeClass,
            finished = false, //一轮游戏结束转动标志 
            step = 0, //一轮游戏的计步数
            cycle = 0, //当前转动圈数   
            cycleTotal = 10, //一轮游戏的总转动圈数
            index = 0, //当前格
            indexPrev, //前一格
            indexStop = 0, //决定在哪一格停止
            indexSpeedUp = 5, //决定在哪一格加速
            indexSpeedDown = 0, //决定在哪一格减速
            timerId1, //定时器对象，开始
            timerId2, //定时器对象，结束
            that = this;
        that.status = STATUS_STOPPED

        // 1. reset config
        this._resetStart = function () {
            that.status = STATUS_STARTING
            lottoArr.forEach(function(el) {
                el.classList.remove(activeClass)  
            })
            indexSpeedUp = 5
        }
        this._resetStop = function () {
            that.status = STATUS_STOPPING
            finished = false
            cycle = 0
            cycleTotal = 3
            indexSpeedUp = 0
        }

        // 2. exec animation && callback
        this._start = function () {
            timerId1 = setInterval(move1, SPEED_LOW);
            function move1() {
                //走N格开始加速
                if (step == indexSpeedUp) {
                    clearInterval(timerId1);
                    timerId1 = setInterval(move1, SPEED_HIGH);
                }

                if (that.status === STATUS_STOPPING){
                    clearInterval(timerId1);
                }

                if (index >= lottoArr.length) {
                    index = 0
                }

                indexPrev = index > 0 ? index - 1 : lottoArr.length - 1
                lottoArr[indexPrev].classList.remove(activeClass)
                lottoArr[index].classList.add(activeClass)

                index++
                step++
            }
        } 
        this._stop = function (stopIndex, cb, cbCtx) {
            stopIndex = stopIndex || 0
            cb = cb || function(){}

            timerId2 = setInterval(move2, SPEED_HIGH);
            function move2() {   
                //跑马灯变速
                if (finished == false) {
                    //跑N圈减速
                    if (cycle > cycleTotal && index == 4) {
                        clearInterval(timerId2);
                        finished = true; //触发结束  
                        timerId2 = setInterval(move2, SPEED_LOW);
                    }
                }

                if (index >= lottoArr.length) {
                    index = 0;
                    cycle++;
                }

                indexPrev = index > 0 ? index - 1 : lottoArr.length - 1;
                lottoArr[index].classList.add(activeClass)
                lottoArr[indexPrev].classList.remove(activeClass)

                //结束转动并选中号码
                if (finished == true && index == stopIndex) {
                    clearInterval(timerId2)
                    step = 0
                    that.status = STATUS_STOPPED
                    cb.call(cbCtx)
                }

                index++
                step++
            }
        }
    }

    Lottery.prototype = {
        constructor: Lottery,
        /**
         * 开始抽奖
         */
        start: function(){
            if(this.status !== STATUS_STOPPED){
                return;
            }
            this._resetStart()
            this._start()
        },
        /**
         * 停止抽奖
         * 
         * @param  {number}   index 表示动画停在哪个奖品序号，奖品序号按从0开始左上角顺时针排列
         * @param  {Function} cb    动画停止后执行的回调函数
         * @param  {*}        ctx   回调函数的上下文
         */
        stop: function(index, cb, ctx){
            if(this.status !== STATUS_STARTING){
                return;
            }
            this._resetStop()
            this._stop(index, cb, ctx)
        }
    };

    return Lottery;
});

static/js/site.js
$(function(){
  $('.highlight').on('click', '.code-toggle', function() {
    var pre = $(this).parents('.highlight')
    if (pre.hasClass('collapse')) {
      pre.removeClass('collapse')
      $(this).text('收起')
    } else {
      pre.addClass('collapse')
      $(this).text('展开')
    }
    return false
  });

  $('h4 em, h3 em, h3 code, h4 code').parent().addClass('doc-api');
  // 给 iframe 加链接
  $('.nico-iframe').each(function(i, item) {
    var src = $(item).find('iframe').attr('src')
    var html = '<a class="new-window" target="_blank" href="' + src + '">新窗口打开</a>'
    $(item).append(html)
  });
  // 给 code 加收起和展开
  $('.highlight').each(function(i, item) {
    var $item = $(item)
    if ($item.height() > 400) {
      $item.append('<a class="code-toggle" href="#">展开</a>')
      $item.addClass('collapse')
    }
  });
});


src/fruit-slot.js
/**
 * 经典水果机抽奖 JS 组件
 *
 * 用于9宫格/12宫格等格子似的抽奖表现形式
 *
 * @version 2.1.0 2016/1/21
 * @author https://github.com/ufologist/fruit-slot
 * @author https://github.com/duowan/lottery
 * @license MIT
 */
(function webpackUniversalModuleDefinition(root, factory) {
    // [UMD](https://github.com/umdjs/umd)
    if(typeof exports === 'object' && typeof module === 'object') {
        module.exports = factory();
    } else if(typeof define === 'function' && define.amd) {
        define([], factory);
    } else if(typeof exports === 'object') {
        exports['FruitSlot'] = factory();
    } else {
        root['FruitSlot'] = factory();
    }
})(this, function() {
    var STATUS_STARTING = 1, STATUS_STOPPING = 2, STATUS_STOPPED = 3,
        SPEED_HIGH = 50, SPEED_LOW = 300;

    var defaults = {
        el: "body",
        lottoItem: "[data-lotto-item]",
        activeClass: "lotto__item--hover"
    };

    // utils
    // http://youmightnotneedjquery.com/#extend
    function extend(out) {
        out = out || {};

        for (var i = 1; i < arguments.length; i++) {
            if (!arguments[i])
                continue;

            for (var key in arguments[i]) {
                if (arguments[i].hasOwnProperty(key))
                    out[key] = arguments[i][key];
            }
        }

        return out;
    }
    function _createCycleArr(lottoElements) {
        var arrSrc = [],arrOrder = [],arrDst = [],topMin = 9999,topMax = 0,leftMin = 9999,leftMax = 0,_arrTop = [],_arrBottom = [],_arrLeft = [], _arrRight = [];

        for (var i = 0, length = lottoElements.length; i < length; i++) {
            var lottoElement = lottoElements[i];
            var obj = lottoElement.getBoundingClientRect()
            obj.elem = lottoElement
            arrSrc.push(obj)
        }

        arrSrc.forEach(function(e) {
            topMin = e.top < topMin ? e.top : topMin;
            topMax = e.top > topMax ? e.top : topMax;
            leftMin = e.left < leftMin ? e.left : leftMin;
            leftMax = e.left > leftMax ? e.left : leftMax;
        })
        arrSrc.forEach(function(e) {
            e.top === topMin && _arrTop.push(e)
            e.top === topMax && _arrBottom.push(e)
            e.left === leftMin && _arrLeft.push(e)
            e.left === leftMax && _arrRight.push(e)
        })
        _arrTop.sort(function(a, b) {
            return a.left - b.left
        })
        _arrRight.sort(function(a, b) {
            return a.top - b.top
        })
        _arrBottom.sort(function(a, b) {
            return b.left - a.left
        })
        _arrLeft.sort(function(a, b) {
            return b.top - a.top
        })
        _arrTop = _arrTop.slice(0, _arrTop.length-1)
        _arrRight = _arrRight.slice(0, _arrRight.length-1)
        _arrBottom = _arrBottom.slice(0, _arrBottom.length-1)
        _arrLeft = _arrLeft.slice(0, _arrLeft.length-1)
        arrOrder = _arrTop.concat(_arrRight.concat(_arrBottom.concat(_arrLeft)))
        arrOrder.forEach(function(e) {
            arrDst.push(e.elem)
        })
        return arrDst
    }

    function FruitSlot(opts) {
        // 0. setup config
        var options = this.options = extend({}, defaults, opts)
        var arr = document.querySelector(options.el).querySelectorAll(options.lottoItem),
            // 格子们默认顺时针排好队,
            // 也可以直接传入奖品元素列表(那么顺序完全由用户来控制)
            // 例如自己按逆时针放置好所有元素, 也可以支持双横排或者双竖排这样的特殊排列布局
            // 这样就可以随心所欲的布局和控制奖品元素的动画了
            // 思考出这种模式主要是受到 https://github.com/areyouse7en/lottery/blob/master/js/lottery.marquee.js
            // 的启发, 还可以参考其基于缓动函数来实现动画的逻辑
            lottoArr = options.lottoArr ? options.lottoArr : _createCycleArr(arr),
            activeClass = options.activeClass,
            finished = false, //一轮游戏结束转动标志 
            step = 0, //一轮游戏的计步数
            cycle = 0, //当前转动圈数   
            cycleTotal = 10, //一轮游戏的总转动圈数
            index = 0, //当前格
            indexPrev, //前一格
            indexStop = 0, //决定在哪一格停止
            indexSpeedUp = 5, //决定在哪一格加速
            indexSpeedDown = 0, //决定在哪一格减速
            timerId1, //定时器对象，开始
            timerId2, //定时器对象，结束
            that = this;
        that.status = STATUS_STOPPED

        // 1. reset config
        this._resetStart = function () {
            that.status = STATUS_STARTING
            lottoArr.forEach(function(el) {
                el.classList.remove(activeClass)  
            })
            indexSpeedUp = 5
        }
        this._resetStop = function () {
            that.status = STATUS_STOPPING
            finished = false
            cycle = 0
            cycleTotal = 3
            indexSpeedUp = 0
        }

        // 2. exec animation && callback
        this._start = function () {
            timerId1 = setInterval(move1, SPEED_LOW);
            function move1() {
                //走N格开始加速
                if (step == indexSpeedUp) {
                    clearInterval(timerId1);
                    timerId1 = setInterval(move1, SPEED_HIGH);
                }

                if (that.status === STATUS_STOPPING){
                    clearInterval(timerId1);
                }

                if (index >= lottoArr.length) {
                    index = 0
                }

                indexPrev = index > 0 ? index - 1 : lottoArr.length - 1
                lottoArr[indexPrev].classList.remove(activeClass)
                lottoArr[index].classList.add(activeClass)

                index++
                step++
            }
        } 
        this._stop = function (stopIndex, cb, cbCtx) {
            stopIndex = stopIndex || 0
            cb = cb || function(){}

            timerId2 = setInterval(move2, SPEED_HIGH);
            function move2() {   
                //跑马灯变速
                if (finished == false) {
                    //跑N圈减速
                    if (cycle > cycleTotal && index == 4) {
                        clearInterval(timerId2);
                        finished = true; //触发结束  
                        timerId2 = setInterval(move2, SPEED_LOW);
                    }
                }

                if (index >= lottoArr.length) {
                    index = 0;
                    cycle++;
                }

                indexPrev = index > 0 ? index - 1 : lottoArr.length - 1;
                lottoArr[index].classList.add(activeClass)
                lottoArr[indexPrev].classList.remove(activeClass)

                //结束转动并选中号码
                if (finished == true && index == stopIndex) {
                    clearInterval(timerId2)
                    step = 0
                    that.status = STATUS_STOPPED
                    cb.call(cbCtx)
                }

                index++
                step++
            }
        }
    }

    FruitSlot.prototype = {
        constructor: FruitSlot,
        /**
         * 开始抽奖
         */
        start: function(){
            if(this.status !== STATUS_STOPPED){
                return;
            }
            this._resetStart()
            this._start()
        },
        /**
         * 停止抽奖
         * 
         * @param  {number}   index 表示动画停在哪个奖品序号，奖品序号从0开始以左上角顺时针排列
         * @param  {Function} cb    动画停止后执行的回调函数
         * @param  {*}        ctx   回调函数的上下文
         */
        stop: function(index, cb, ctx){
            if(this.status !== STATUS_STARTING){
                return;
            }
            this._resetStop()
            this._stop(index, cb, ctx)
        }
    };

    return FruitSlot;
});